### ISSUE TITLE
Courses - master issue

### ISSUE THREAD

**Issue by rogernclarke** [rogernclarke/Rankings#60]
Add courses functionality. Precise functions to be suggested / decided on this issue - each function to be built to then have its own issue created.

Suggested functionality:
- Course records (since 2017), male and female
- Links to each event held on the course

**Comment by drnickw** [comment]
I`ve been doing  a bit more work on this. I think i`ll need a field on the event table which will be a "weighted average speed" to set an average for the event itself.
This may be enough depending on what processing you can do. 
If we can show all the results on this course and allow sorts by time/cat/gender/race date etc then I think people will be able to get the info they want, and comparing the different "w_a_s" should show the relative float for each event. this might avoid setting up a new "courses" table, unless there is other data we want to hold.

**Comment by drnickw** [comment]
so, ive got a load of new fields to pass, as follows
1, weighted average speed - the ave spd of a notional rider ranked 2000 (r2000) - split into 2 fields, one is the R2000 for the event, the other for the course (all events)
2. record time (spindata data only)
3. rank needed to achieve that time - currently r25 is the toughest record to get with a -14, and johnAs Buxton in second with - 12.
4, the time you would have got on  the day if it had been an average day
5, the time you would have got on  the day if it had been the best day recorded.

first 3 would go on event table and last 2 on ride, unless you want to make a courses table for 1b,2 and 3, which would also need course code and distance.
one consideration is whether we want to update everything every time. 1a 4 and 5 could be static based on data to that point or updated in  light of new events on course subsequently.

Can you think of any other data you`d like from me re this? looking back i should probably do a female CR with rank needed. do you think field strength would be interesting - im not sure now as it doesnt affect the rankings and is obvious from the listing.

**Comment by rogernclarke** [comment]
We should definitely be creating a course table.1b would go on the course table.I don't think we need to store 2 as it is a simple SQL query to pick up from the rides table although I suppose we need to allow for short events that are nominally on the same course. Any suggestions how best to handle those? An indicator on the event or each ride in a shortened event?Is weighted average speed for r2000 going to be the most easily understood measure of how fast an event was? Or would a factor or percentage of the best day for r2000 be better?Is rank needed going to be the rank of the rider when they set the course record? In which case do we have that on the rides table?4&5 is this by taking a rider's score for their ride in that event and factoring in how hard the day was compared with best and average events? So we only have the capability of doing this for riders who took part in that event using their score on that day?Just trying to think through bigger picture where else we might want to go with this. What if I am an r3000 rider who hasn't ridden a course and want to know what sort of time I could get in a best or average day? I could look at results for an event, find a rider who scored close to 3000, then look at what time they would have had for an average of best day. Maybe having 120 random data points from those who rode is ok for that, but also wondering if having a standard set of scoring data points might be useful e.g. 1,50,100,250,500 etc.Definitely think 1a should be kept up to date, so probably also 4 and 5 but that could be a lot of updates each week!On 8 Nov 2019 17:46, drnickw <notifications@github.com> wrote:so, ive got a load of new fields to pass, as follows
1, weighted average speed - the ave spd of a notional rider ranked 2000 (r2000) - split into 2 fields, one is the R2000 for the event, the other for the course (all events)
2. record time (spindata data only)
3. rank needed to achieve that time - currently r25 is the toughest record to get with a -14, and johnAs Buxton in second with - 12.
4, the time you would have got on  the day if it had been an average day
5, the time you would have got on  the day if it had been the best day recorded.
first 3 would go on event table and last 2 on ride, unless you want to make a courses table for 1b,2 and 3, which would also need course code and distance.
one consideration is whether we want to update everything every time. 1a 4 and 5 could be static based on data to that point or updated in  light of new events on course subsequently.
Can you think of any other data you`d like from me re this? looking back i should probably do a female CR with rank needed. do you think field strength would be interesting - im not sure now as it doesnt affect the rankings and is obvious from the listing.

**Comment by drnickw** [comment]
I knew youd want a course table, just testing!

I screen all events to make sure they are at the correct distance, by
checking they arent too far off the norm for the course. Whilst this isnt
100% it picks up anything significant.
When I find these events I change the distance so thats why I suggest the
course table has course code and distance as the keys.
Whilst it may be meaning less to have a short course with a record etc at
least it keeps it separate from the standard course.

An R2000 as a speed allows courses to be compared with each other, as well
as within events on that course.
I can do a percentage of standard as well if it helps, but 1500 of the
approx 2500 are within 1% of the average for the course, and only 1% are
outside 5% and they are mostly funnies (eg roadbikes only) or courses with
only 2 rides and very different conditions.
We could make up an arbitrary figure like 10x the percentage difference
from the average conditions so a range would usually be -10 to +10 with
extremes going out to -150 and +60 (where + is better, (there is more scope
for an event to be very slow than very fast))
It might be more engaging but would be less meaningful.
Having the times for average conditions and float conditions on the result
sheet against each riders result would negate people needing to try and do
sums to see how they would have performed.

Rank for record is transmuted to the rank needed on an average day to set
the time so it takes the condition variability out of the record ride -
makes it quite interesting too, as it shows how good a performance it
actually was.

4 and 5 can become columns on the results table (display to be considered).

Its actually very easy to keep 4+5 updated, I have to run through all
events on a course when there is a new event anyway to reset the average
and only need to do this for courses run that week so its no big deal.
It just means the historical data is shifting but probably thats better.

The R3000 rider is an issue. I have already generated R1000 to R8000 for
each event (although they are not always available, think juvenile event or
nat 25). You can display in bands like this.
I cant reliably produce a formula for getting the time for a rank as the
curve does vary a bit.
I could do a standard curve multiplier table for you to generate but it
might not map perfectly to the results so would look odd as you might get a
result which wasnt in line with the actual results on the day and isnt at
all reliable above 5000.
I can do 1:max rank for every event so you can enter your rank and it tells
you exactly but this would mean huge files to cover every rank, or we could
do 1 for each course/distance - theres currently 662 and max rank is about
10k so thered be 6,620,000 times in there as opposed to over 25,000,000
times if we did it for every event. Looking at the vets standard table even
the first option would be 29 times as big so I dont think thats realistic,
unless I`m missing something. If we did it in bands we can reduce this by
any factor you like - could we do it in bands of 100 and then use these
approx 100 points (ie, max rank/100) to generate a graph of time against
rank for the standard conditions on a course?  Would it then be possible to
move a line over the graphs axis to get down to individual ranks or times
by just assuming a straight line between each 100 ranks.
I guess you could do this with a simple data entry box for a time or rank
and get the corresponding result and wouldnt need a graph but the graph
would look nice and Im keen to promote our graphs as the fundamental basis
for the ranking - I was a bit alarmed that someone suggested  CTT should be
looking at developing an ELO system to replace their rankings. I`m
confident we could prove the value of our system against elo by our
validation via predictions but i suspect we wouldnt get a look in if CTT
developed an acceptable (but not quite as good as Spindata) official
rankings.

**Comment by rogernclarke** [comment]
Agree we need to keep a separate record for each course code / distance combination. I need to think through further if we key that on course code and distance or create a unique id. There will be pros and cons of each.

For each course we should add an indicator to show which is the proper course version. We could if we want choose to only display course records for the proper course version.

I'd forgotten we were also comparing one course to another so it does need to be speed.

I quite like the +/- idea for good and bad days.

R3000 - without deciding exactly what we will do with this it is enough to know we don't want an R2000 column but a further table with one record per ranking data point per course. A graph of standard data points makes sense.

**Comment by rogernclarke** [comment]
I've started on this but getting a bit lost in what we need. So far I've got:
course_id
course_code
distance
official_version_ind (to indicate if the version of the course for which a record should be held - maybe needs a better name)
weighted_mph
comments

We've then got to deal with course records and rankings required to achieve them. At least for men and women but no reason not to also hold records for other categories? This then implies a course_records table keyed by course_id and category, holding a time and ranking required for each?

And then are we doing weighted average speeds for other ranking points in a separate table?

**Comment by drnickw** [comment]
1." I've started on this but getting a bit lost in what we need. So far I've got:
 course_id
 course_code
 distance
 official_version_ind (to indicate if the version of the course for which a record should be held - 
 maybe needs a better name)
 weighted_mph
 comments"

Yes, I think this is enough for now.

2."We've then got to deal with course records and rankings required to achieve them. At least for men and women but no reason not to also hold records for other categories? This then implies a course_records table keyed by course_id and category, holding a time and ranking required for each?"
"I don't think we need to store 2 as it is a simple SQL query to pick up from the rides table 
I would suggest we generate the records with SQL"
"And then are we doing weighted average speeds for other ranking points in a separate table?"

I agree, we can generate the records from SQL then we dont need to decide and fix the records we want. The rank for the time is easy if we have a table relating rank to time for each course/distance combo as discussed above. Once you know the record time you want you just look up the rank. This also enables us to generate a graph to see what you would do on a course with a particular rank. Looking at the data, as above, a time for every rank would be too much, but dividing the ranks by 100 is not very accurate for the high numbers. I`m incline to go with 10s, so a time for rank 1 and a time for rank 11 etc. This would necessitate a calculation if you want a time to the unit - probably worth doing for the records but not for the lookup/graph. As the field is diminishing over time I think we need to fix the ranks at 10000 and then for each current rank adjust for the total number of currently active riders as coming in at rank 5000 now is bottom half but 2 years ago was top half.

3. "And then are we doing weighted average speeds for other ranking points in a separate table?"
See 2!

So, in summary a course table along the lines you suggest. A rank to time table with 1000 ranks (in 10s) and a time for every one for each course/distance combo (currently < 700). 2 new columns on the event table, one for weighted MPH and one for percent from the norm for this course (unless you want to deduce this from the course table via SQL?) . 2 new fields on the ride table being time on average day for this course and time on best day for this course (for each rider)

**Comment by rogernclarke** [comment]
Ok, think I'm making progress on this.

New table added called tt_course_rank_times, keyed on course_id and rank, with a column for time.
Question - what about 12 and 24 hour courses? Do we need a distance column as well?

Events table - I've added weighted_mph as a new column. Think I'm happy to calculate % from norm in the SQL as it is a simple join for each event to course table.

Rides table - I'm assuming you are going to populate these fields, on loading startsheet?
Question - if we have values for average and best days available per ride, does the course_rank_times table just hold data for an average day on the course rather than a best day on the course?

If you are going in ranking steps of 10 then could I suggest going 1, 10, 20, 30 etc. rather than 1, 11, 21 as it is going to look a little odd on the graph axis.

And lastly, for now, moving to a course_id for the primary key of the courses table does raise the question of joining an event to the course. Joining on course code and distance is a little messy. Should we bite the bullet and allocate the new course id to past events and replace course code and distance in the event table with the new course id?

**Comment by drnickw** [comment]
Yes, we can have a distance column too if you like - assume it will take up very little extra memory as hardly ever filled?

I was going to magnify the percent by 10 and give it as an integer so we get a reasonable range. It`s moving away from a pure understandable figure to an easier one to work with - will be in the -50 to + 15 sort of category rather than -5.0 to + 1.5 but it doesnt really matter. Danger of the pure % is people using it to work out their time adjustment - its not consistent across the whole field.

Rides - on loading SS the average will be the prediction. I`m thinking these fields will be to show what the performance you actually did (as opposed to your pre rank) would be worth so wasnt proposing giving them in advance. Could give a notional_best_time_from_pre_rank on but it may be beaten anyway if the day is a new fastest so not sure I want to put this data forward. 

I`d like the average day to be the exclusive one as the fastest is based on a single event so much more likely to be wrong.  We are already offering % from norm and record times which I think is enough. I only really wanted the best possible as an option column on the results out of interest. I think the thing people will really want is the difference in time between what they were predicted to do and what their effort on the day would have been worth if it had been an average day (using the pre race data) so a sort of "value added/lost" of their ride.

1,10,20 is fine. Only thing is are you going to offer a time for a specific rank or just a graph. If we want to give individuals a time for their rank in future as a field on their profile we`ll need to do the maths between the 10s. Will just need an exception for the top 10 but will hardly ever be required.

Course code - i guess thats the "right" way to do it, if you think its worth making the change - its easy for me. Presumably will need an extra look up for your retrievals.

**Comment by rogernclarke** [comment]
The decimal field for holding distances takes the space whether it stores anything in it or not. So we are going to have several columns on the rides table that take space that mostly hold nothing. I think we're ok with that for the moment a long as it doesn't get out of hand.

I really hadn't thought of what we're doing to display this data, so no plans of graph vs specifics.

The dev database now has all the changes I think are needed. Would be worth you having a look over it. I've changed the name of tt_course_rank_times to tt_course_rank_performances to reflect that it also holds distances. I'm now having second thoughts that maybe there should be a tt_course_rank_times table and a separate tt_course_rank_distances table!

Let me know if you spot anything else.

Course id has been added to events table. Once you have populated it for all events I'll update the app code to use it.

**Comment by drnickw** [comment]
"Events table - I've added weighted_mph as a new column. Think I'm happy to calculate % from norm in the SQL as it is a simple join for each event to course table."

In the end I`ve changed the data a bit! The R2000 average speed works for comparing different courses as theres no other way, but for events on a specific course its much better to take an average of the percentage offset of everyones ride (well actually I`m using the most accurate 85% as theres no benefit in keeping the outliers) and use this as the adjustment for the event. This gives far more reliable data than assuming the offset at 2000 is right for the whole event. Also turns out the percentage change across the rankings actually maps quite well so its as valid for fast and slow riders.
This does mean, however, we`ll need a % figure (or a +/- variance from 100% ?x10) to represent variations within a course on the table.

(or a +/- variance from 100% ?x10) - really not sure how best to display this. percent currently varies from 95.886 (ie everyone was over 4% faster on average) to 108.692 although there are some more distant outliers which I`ve excluded for various reasons, which stretch the field to 92 and 110. of the near 2500 events 500 have an offset of between 100 and 101. We could just give the straight percent to 1 decimal place or x10 to have 1000 as the norm and a range of 959 to 1087 or an offset from 100 so -41 to +87 or of course -4.1 to +8.7, bearing in mind 1/5th of the events will be 0.0 to 1.0, unless you can think of a better way to present this?  Ultimately it doesnt matter at this stage as I can give you the cold hard percentage and you can show whatever we think is best.
I`ll need to recalculate everything for each event run so the data will change with time too, and Im adjusting for date to take the overall field size in too, as this affects the ranking given on the day ( a rank of 1000 in 2017 is better than a rank of 1000 now as theres fewer riders now)

**Comment by rogernclarke** [comment]
I think a plus or minus figure might be easiest to understand, but I'd probably go with + for a faster ride as people tend to think about their time or speed being better. So +4.1 to -8.7?

What format do you think the number should be stored in MySQL? Not sure if better to have the full percentage number or just as it will be displayed. Once i know that and the level of precision I can set the field up for you.

**Comment by drnickw** [comment]
I`m actually working on this just now. The field you have currently on the
event table is "weigthed mph" but I`m not going to give this for events,
just an offset percent. weighted mph is relevant for courses still, however.

What I thought I`d discovered is there is a discrepancy across the ranks,
so the slower riders were beating their prediction by more. When I tried to
correct by rider for the conditions on the day it looked like it all came
back into line so I didnt follow this up.

However what is really interesting is looking at the offset for
separate races on the same course on the same day.
Overall they tally really well, even though they are processed
separately which validates the method. Obviously if its a 72inch or
roadbikes event its much worse than the TT event which is right.
However it appears when there are different classes of rider on the same
event eg a womens or vets or junior, then 9 time out of 10 the juniors
outperform more than the women who outperform more than the vets who
outperform more than the seniors.

I need to do a bit more work on this now as this would again suggest there
is a noticable difference between the over/underperformance based on rank,
even taking condition into account.

In terms of accuracy out of 640 events which are paired (or more)  half
have a variation of < 0.2 (ie if one is 1.3 the other will be between 1.1
and 1.5)  another 140 will have variation < 0.6. The others are big
outliers eg road vs tt which you would expect. So i dont think theres any
point having an accuracy above 1 dec place as its not valid beyond this
level.

Not much point in you processing as you can always work it backwards if you
want. I may as well give you the display figure ie +4.1 to -8.7.

**Comment by rogernclarke** [comment]
It was just a quick change so I've changed the weighted_mph column on the events table to be offset_percentage and reduced the size to decimal 3,1
